packages/javascript-decorators/src/tests/memoize.test.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import memoize, {MAX_MAP_ENTRIES} from '../memoize';
+
+describe('memoize()', () => {
+  describe('client', () => {
+    it('recalculate the result when first argument changed', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({addOneSpy: spy});
+
+      expect(testClass.addOne(1)).toEqual(2);
+      expect(testClass.addOne(2)).toEqual(3);
+      expect(spy).toHaveBeenCalledTimes(2);
+    });
+
+    it('gets the result from cache when the first argument stay the same', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({addOneSpy: spy});
+
+      expect(testClass.addOne(1)).toEqual(2);
+      expect(testClass.addOne(1)).toEqual(2);
+      expect(spy).toHaveBeenCalledTimes(1);
+    });
+
+    it('does not put the result into map cache, if map cache is over its max limit', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({addOneSpy: spy});
+
+      for (let i = 0; i < MAX_MAP_ENTRIES; i++) {
+        expect(testClass.addOne(i)).toEqual(i + 1);
+      }
+
+      // 0 is in the cache
+      expect(testClass.addOne(0)).toEqual(1);
+      expect(testClass.addOne(1)).toEqual(2);
+      expect(spy).toHaveBeenCalledTimes(MAX_MAP_ENTRIES);
+
+      expect(testClass.addOne(MAX_MAP_ENTRIES)).toEqual(MAX_MAP_ENTRIES + 1);
+      expect(spy).toHaveBeenCalledTimes(MAX_MAP_ENTRIES + 1);
+
+      // 0 is no longer in the cache
+      expect(testClass.addOne(0)).toEqual(1);
+      expect(spy).toHaveBeenCalledTimes(MAX_MAP_ENTRIES + 2);
+    });
+  });
+
+  describe('only argument is object', () => {
+    it('recalculate the result when the first argument changed', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({getValuesSpy: spy});
+
+      expect(testClass.getValues({one: 1, two: 2})).toEqual([1, 2]);
+      expect(testClass.getValues({one: 3, four: 4})).toEqual([3, 4]);
+      expect(spy).toBeCalledTimes(2);
+    });
+
+    it('gets the result from cache when the first argument stay the same', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({getValuesSpy: spy});
+
+      const testObject1 = {one: 1, two: 2};
+      expect(testClass.getValues(testObject1)).toEqual([1, 2]);
+      expect(testClass.getValues(testObject1)).toEqual([1, 2]);
+      expect(spy).toBeCalledTimes(1);
+    });
+
+    it('does not change the result when the argument was changed in value only', () => {
+      const spy = jest.fn();
+      const testClass = createTestClass({getValuesSpy: spy});
+
+      const testObject1 = {one: 1, two: 2};
+      expect(testClass.getValues(testObject1)).toEqual([1, 2]);
+
+      testObject1.one = 2;
+      expect(testClass.getValues(testObject1)).toEqual([1, 2]);
+
+      expect(spy).toBeCalledTimes(1);
+    });
+  });
+
+  describe('customized resolver', () => {
+    it('recalculate the result when the resolver result was changed', () => {
```

Comment:
This is the only block that uses `getName`, so I'd prefer to see a specific `class ResolverTester` defined in this block.

This fits with our convention of keeping test helpers isolated, and not overloading them with many methods.

Reply:
Looking at the other `describe` blocks, it looks like we could have an isolated tester class for those, too.

Reply:
I isolated them to a create class method that gets call.
which is a pattern I have seem before as well (for things like `graphqlClient`)
if there more worries for class that should not be create like this?

Reply:
Having a test class with _every_ type of memoize usage is convenient for the writer of the test, but inconvenient for new readers.  Before I can understand any test, I have to jump down to the helper class, and understand how every `@memoize` method works :/

If each `describe` block defines a test helper, I can focus on that, and not jump around so much.

Reply:
I strongly agree, repetition feels like it would be very useful here